File: sites/stat/legacy/build-stat.mjs
#!/usr/bin/env node
/**
 * Парсит chat.txt и генерирует stat.html с встроенными сообщениями.
 * Запуск: node build-stat.mjs
 */
import { readFileSync, writeFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const ROOT = dirname(fileURLToPath(import.meta.url));
const chatText = readFileSync(join(ROOT, 'chat.txt'), 'utf8');
const aliases = JSON.parse(readFileSync(join(ROOT, 'name-aliases.json'), 'utf8'));

const HEADER_RE = /^\[(\d{2})\.(\d{2})\.(\d{4})\s+(\d{2}):(\d{2})\]\s+(.+?):\s*$/;

function normalizeAuthor(raw) {
  if (raw.startsWith('***')) return 'Система';
  return aliases.aliases[raw] ?? raw;
}

function parseChat(text) {
  const lines = text.split('\n');
  const messages = [];
  let i = 0;
  let id = 0;

  while (i < lines.length) {
    const line = lines[i];
    const m = line.match(HEADER_RE);
    if (!m) {
      i++;
      continue;
    }
    const [, dd, mm, yyyy, hh, min, authorRaw] = m;
    const bodyLines = [];
    i++;
    while (i < lines.length && !lines[i].match(HEADER_RE)) {
      if (lines[i] !== '' || bodyLines.length > 0) bodyLines.push(lines[i]);
      i++;
    }
    while (bodyLines.length > 0 && bodyLines[bodyLines.length - 1] === '') {
      bodyLines.pop();
    }
    const textBody = bodyLines.join('\n');
    id++;
    messages.push({
      id,
      datetime: `${dd}.${mm}.${yyyy} ${hh}:${min}`,
      date: `${dd}.${mm}.${yyyy}`,
      year: +yyyy,
      month: +mm,
      day: +dd,
      authorRaw,
      author: normalizeAuthor(authorRaw),
      text: textBody,
      isSystem: authorRaw.startsWith('***'),
    });
  }
  return messages;
}

function isRelevantWindow(msg) {
  const { year, month, day } = msg;
  if (year !== 2026) return false;
  if (month === 5 && day >= 25) return true;
  if (month === 6 && day <= 22) return true;
  return false;
}

function isMediaOnly(text) {
  const t = text.trim();
  return /^\[(видеосообщение|файл|фото|стикер|голосовое сообщение)([^\]]*)\]$/i.test(t);
}

function hasNumberedHabits(text) {
  const lines = text.split('\n').slice(0, 20);
  let numbered = 0;
  for (const line of lines) {
    if (/^\s*\d+[\.\)]\s+\S/.test(line)) numbered++;
  }
  return numbered >= 2;
}

function classify(msg) {
  if (msg.isSystem || isMediaOnly(msg.text)) return 'simple';

  const text = msg.text;
  const lower = text.toLowerCase();
  const firstLines = text.split('\n').slice(0, 5).join('\n').toLowerCase();
  const intro = lower.slice(0, 400);

  if (/\bманифест\b/.test(intro) && !/отч[её]т\s*(за\s+)?\d/i.test(firstLines)) {
    return 'manifest';
  }
  if (/мои\s+цели\s+на\s+марафон/.test(intro) && !/\bотч[её]т\b/i.test(firstLines)) {
    return 'manifest';
  }

  const reportMarkers = [
    /\bотч[её]т\b/,
    /\bмой\s+отч[её]т\b/,
    /отч[её]т\s+за\s/,
    /отч[её]т\s+\d/,
    /мои\s+успехи\s+за\s+\d/,
    /^отч[её]т\s+\d/i,
  ];
  const hasReportMarker = reportMarkers.some((re) => re.test(firstLines) || re.test(lower.slice(0, 200)));
  const hasCheckmarks = /[✅❌🟡]/.test(text);
  const hasDoneWords = /\b(сделал[аи]?|не\s+сделал[аи]?|не\s+выполнен[оа]?|выполнен[оа]?\s+с\s+опозданием)\b/i.test(text);
  const hasReportDate = /отч[её]т\s*(за\s*)?\d{1,2}[\.\/_]\s*\d{1,2}/i.test(firstLines);

  if (hasReportMarker || (hasCheckmarks && hasNumberedHabits(text)) || (hasReportDate && hasNumberedHabits(text))) {
    if (!/^(мои\s+(цели|привычки|планы))/i.test(firstLines.trim()) || /\bотч[её]т\b/i.test(firstLines)) {
      return 'report';
    }
  }
  if (hasCheckmarks && /\(\d+\/\d+\)/.test(text)) return 'report';
  if (hasDoneWords && hasNumberedHabits(text) && /\d{1,2}[\.\/_]\d{1,2}/.test(firstLines)) return 'report';

  const manifestMarkers = [
    /\bманифест\b/,
    /мои\s+привычки/,
    /мои\s+планы\s+на/,
    /мои\s+цели\s+на\s+(марафон|июн|91)/,
    /мои\s+цели\s*-/,
    /формирую\s+следующие\s+привычки/,
    /привычки,?\s+которые\s+я\s+беру/,
    /мои\s+цели\s+на\s+91\s+день/,
  ];
  const isManifestKeyword = manifestMarkers.some((re) => re.test(lower));
  const isEarlyCycle =
    (msg.month === 5 && msg.day >= 28) ||
    (msg.month === 6 && msg.day <= 3);

  if (isManifestKeyword) {
    const looksLikeReport = /\bотч[её]т\b/i.test(firstLines) && !/^↩/.test(text);
    if (!looksLikeReport) return 'manifest';
  }
  if (hasNumberedHabits(text) && isEarlyCycle && !hasCheckmarks && !/\bотч[её]т\b/i.test(lower.slice(0, 100))) {
    if (/мои\s+цел/i.test(lower) || /привычк/i.test(lower) || /марафон/i.test(lower)) {
      return 'manifest';
    }
  }

  if (/^благодарю[:\s]/im.test(text) && !hasCheckmarks && !hasNumberedHabits(text)) {
    return 'simple';
  }

  return 'simple';
}

const messages = parseChat(chatText).map((msg) => ({
  ...msg,
  opinion: classify(msg),
  inWindow: isRelevantWindow(msg),
}));

const template = readFileSync(join(ROOT, 'stat.template.html'), 'utf8');
const html = template.replace('/*__MESSAGES__*/', JSON.stringify(messages));
writeFileSync(join(ROOT, 'stat.html'), html, 'utf8');

const counts = { simple: 0, manifest: 0, report: 0 };
for (const m of messages) counts[m.opinion]++;
console.log(`stat.html: ${messages.length} сообщений`);
console.log(`Мнение: простое=${counts.simple}, манифест=${counts.manifest}, отчёт=${counts.report}`);
console.log(`В окне мая–22.06: ${messages.filter((m) => m.inWindow).length}`);

File: sites/stat/legacy/build-stat.py
#!/usr/bin/env python3
"""Парсит chat.txt и генерирует stat.html с встроенными сообщениями."""
import json
import re
from pathlib import Path

ROOT = Path(__file__).parent
HEADER_RE = re.compile(r"^\[(\d{2})\.(\d{2})\.(\d{4})\s+(\d{2}):(\d{2})\]\s+(.+?):\s*$")
MEDIA_ONLY_RE = re.compile(
    r"^\[(видеосообщение|файл|фото|стикер|голосовое сообщение)([^\]]*)\]$",
    re.I,
)


def load_aliases():
    data = json.loads((ROOT / "name-aliases.json").read_text(encoding="utf-8"))
    return data["aliases"]


def normalize_author(raw, aliases):
    if raw.startswith("***"):
        return "Система"
    return aliases.get(raw, raw)


def parse_chat(text, aliases):
    lines = text.split("\n")
    messages = []
    i = 0
    msg_id = 0
    while i < len(lines):
        m = HEADER_RE.match(lines[i])
        if not m:
            i += 1
            continue
        dd, mm, yyyy, hh, mi, author_raw = m.groups()
        body = []
        i += 1
        while i < len(lines) and not HEADER_RE.match(lines[i]):
            if lines[i] != "" or body:
                body.append(lines[i])
            i += 1
        while body and body[-1] == "":
            body.pop()
        msg_id += 1
        messages.append(
            {
                "id": msg_id,
                "datetime": f"{dd}.{mm}.{yyyy} {hh}:{mi}",
                "date": f"{dd}.{mm}.{yyyy}",
                "year": int(yyyy),
                "month": int(mm),
                "day": int(dd),
                "authorRaw": author_raw,
                "author": normalize_author(author_raw, aliases),
                "text": "\n".join(body),
                "isSystem": author_raw.startswith("***"),
            }
        )
    return messages


def is_relevant_window(msg):
    y, mo, d = msg["year"], msg["month"], msg["day"]
    if y != 2026:
        return False
    if mo == 5 and d >= 25:
        return True
    if mo == 6 and d <= 22:
        return True
    return False


def is_media_only(text):
    return bool(MEDIA_ONLY_RE.match(text.strip()))


def has_numbered_habits(text):
    numbered = sum(1 for line in text.split("\n")[:20] if re.match(r"^\s*\d+[\.\)]\s+\S", line))
    return numbered >= 2


def classify(msg):
    if msg["isSystem"] or is_media_only(msg["text"]):
        return "simple"

    text = msg["text"]
    lower = text.lower()
    first_lines = "\n".join(text.split("\n")[:5]).lower()
    intro = lower[:400]

    # Явный манифест (слово «манифест» или цели на цикл без шапки отчёта)
    if re.search(r"\bманифест\b", intro) and not re.search(
        r"отч[её]т\s*(за\s+)?\d", first_lines, re.I
    ):
        return "manifest"
    if re.search(r"мои\s+цели\s+на\s+марафон", intro) and not re.search(
        r"\bотч[её]т\b", first_lines, re.I
    ):
        return "manifest"

    report_markers = [
        re.compile(r"\bотч[её]т\b"),
        re.compile(r"\bмой\s+отч[её]т\b"),
        re.compile(r"отч[её]т\s+за\s"),
        re.compile(r"отч[её]т\s+\d"),
        re.compile(r"мои\s+успехи\s+за\s+\d"),
        re.compile(r"^отч[её]т\s+\d", re.I),
    ]
    has_report_marker = any(
        r.search(first_lines) or r.search(lower[:200]) for r in report_markers
    )
    has_checkmarks = bool(re.search(r"[✅❌🟡]", text))
    has_done_words = bool(
        re.search(
            r"\b(сделал[аи]?|не\s+сделал[аи]?|не\s+выполнен[оа]?|выполнен[оа]?\s+с\s+опозданием)\b",
            text,
            re.I,
        )
    )
    has_report_date = bool(
        re.search(r"отч[её]т\s*(за\s*)?\d{1,2}[\.\/_]\s*\d{1,2}", first_lines, re.I)
    )

    if has_report_marker or (
        has_checkmarks and has_numbered_habits(text)
    ) or (has_report_date and has_numbered_habits(text)):
        if not re.match(r"^(мои\s+(цели|привычки|планы))", first_lines.strip()) or re.search(
            r"\bотч[её]т\b", first_lines, re.I
        ):
            return "report"
    if has_checkmarks and re.search(r"\(\d+/\d+\)", text):
        return "report"
    if has_done_words and has_numbered_habits(text) and re.search(
        r"\d{1,2}[\.\/_]\d{1,2}", first_lines
    ):
        return "report"

    manifest_markers = [
        re.compile(r"\bманифест\b"),
        re.compile(r"мои\s+привычки"),
        re.compile(r"мои\s+планы\s+на"),
        re.compile(r"мои\s+цели\s+на\s+(марафон|июн|91)"),
        re.compile(r"мои\s+цели\s*-"),
        re.compile(r"формирую\s+следующие\s+привычки"),
        re.compile(r"привычки,?\s+которые\s+я\s+беру"),
        re.compile(r"мои\s+цели\s+на\s+91\s+день"),
    ]
    is_manifest_keyword = any(r.search(lower) for r in manifest_markers)
    is_early_cycle = (msg["month"] == 5 and msg["day"] >= 28) or (
        msg["month"] == 6 and msg["day"] <= 3
    )

    if is_manifest_keyword:
        looks_like_report = bool(re.search(r"\bотч[её]т\b", first_lines, re.I)) and not text.startswith("↩")
        if not looks_like_report:
            return "manifest"
    if (
        has_numbered_habits(text)
        and is_early_cycle
        and not has_checkmarks
        and not re.search(r"\bотч[её]т\b", lower[:100])
    ):
        if re.search(r"мои\s+цел", lower) or re.search(r"привычк", lower) or re.search(
            r"марафон", lower
        ):
            return "manifest"

    if re.search(r"^благодарю[:\s]", text, re.I | re.M) and not has_checkmarks and not has_numbered_habits(text):
        return "simple"

    return "simple"


def main():
    aliases = load_aliases()
    chat_text = (ROOT / "chat.txt").read_text(encoding="utf-8")
    template = (ROOT / "stat.template.html").read_text(encoding="utf-8")

    messages = []
    for msg in parse_chat(chat_text, aliases):
        msg["opinion"] = classify(msg)
        msg["inWindow"] = is_relevant_window(msg)
        messages.append(msg)

    html = template.replace("/*__MESSAGES__*/", json.dumps(messages, ensure_ascii=False))
    (ROOT / "stat.html").write_text(html, encoding="utf-8")

    counts = {"simple": 0, "manifest": 0, "report": 0}
    for m in messages:
        counts[m["opinion"]] += 1
    print(f"stat.html: {len(messages)} сообщений")
    print(
        f"Мнение: простое={counts['simple']}, манифест={counts['manifest']}, отчёт={counts['report']}"
    )
    print(f"В окне мая–22.06: {sum(1 for m in messages if m['inWindow'])}")


if __name__ == "__main__":
    main()
